Sources/Moho/ScopeCheck.swift

```diff
@@ -0,0 +1,576 @@
+/// ScopeCheck.swift
+///
+/// Copyright 2017, The Silt Language Project.
+///
+/// This project is released under the MIT license, a copy of which is
+/// available in the repository.
+
+import Lithosphere
+
+extension NameBinding {
+  /// Checks that a module, its parameters, and its child declarations are
+  /// well-scoped.  In the process, `NameBinding` will quantify all non-local
+  /// names and return a simplified well-scoped (but not yet type-correct)
+  /// AST as output.
+  ///
+  /// Should scope check detect inconsistencies, it will diagnose terms and
+  /// often drop them from its consideration - but not before binding their
+  /// names into scope to aid recovery.
+  ///
+  /// This pass does not fail thus it is crucial that the diagnostic
+  /// engine be checked before continuing on to the semantic passes.
+  public func scopeCheckModule(_ module: ModuleDeclSyntax) -> DeclaredModule {
+    let params = module.typedParameterList.map(self.scopeCheckTelescope)
+    let filteredDecls = self.withScope(walkNotations(module)) { _ in
+      return self.reparseDecls(module.declList)
+    }
+    return DeclaredModule(
+      moduleName: self.activeScope.nameSpace.module,
+      params: params,
+      namespace: self.activeScope.nameSpace,
+      decls: filteredDecls.flatMap(self.scopeCheckDecl))
+  }
+
+  /// Scope check a declaration that may be found directly under a module.
+  private func scopeCheckDecl(_ syntax: DeclSyntax) -> [Decl] {
+    precondition(!(syntax is FunctionDeclSyntax ||
+      syntax is FunctionClauseDeclSyntax))
+    switch syntax {
+    case let syntax as ModuleDeclSyntax:
+      return [.module(self.scopeCheckModule(syntax))]
+    case let syntax as ReparsedFunctionDeclSyntax:
+      return self.scopeCheckFunctionDecl(syntax)
+    case let syntax as DataDeclSyntax:
+      return self.scopeCheckDataDecl(syntax)
+    case let syntax as RecordDeclSyntax:
+      return self.scopeCheckRecordDecl(syntax)
+      // FIXME: Implement import validation logic.
+//    case let syntax as ImportDeclSyntax:
+//    case let syntax as OpenImportDeclSyntax:
+    default:
+      print(type(of: syntax))
+      fatalError()
+    }
+  }
+
+  /// Scope check and validate an expression.
+  private func scopeCheckExpr(_ syntax: ExprSyntax) -> Expr {
+    switch syntax {
```

Comment:
This function is pretty big -- would be good to split this up into smaller units. Also maybe split this into several files?

`ScopeCheckExpr.swift`, `ScopeCheckDecl.swift`, etc. with extensions?